pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -470,262 +559,611 @@ fn storage_sync_loop<
                         new_timeline_states.len()
                     );
                     // Batch timeline download registration to ensure that the external registration code won't block any running tasks before.
-                    apply_timeline_sync_status_updates(conf, index, new_timeline_states);
+                    apply_timeline_sync_status_updates(conf, &index, new_timeline_states);
                 }
             }
-            LoopStep::Shutdown => {
+            ControlFlow::Break(()) => {
                 info!("Shutdown requested, stopping");
                 break;
             }
         }
     }
 }
 
-async fn loop_step<
-    P: std::fmt::Debug + Send + Sync + 'static,
-    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
->(
+async fn loop_step<P, S>(
     conf: &'static PageServerConf,
-    receiver: &mut UnboundedReceiver<SyncTask>,
-    remote_assets: Arc<(S, RemoteIndex)>,
+    receiver: &mut UnboundedReceiver<(ZTenantTimelineId, SyncTask)>,
+    storage: Arc<S>,
+    index: RemoteIndex,
     max_concurrent_sync: NonZeroUsize,
     max_sync_errors: NonZeroU32,
-) -> LoopStep {
+) -> ControlFlow<(), HashMap<ZTenantId, HashMap<ZTimelineId, TimelineSyncStatusUpdate>>>
+where
+    P: Debug + Send + Sync + 'static,
+    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
+{
     let max_concurrent_sync = max_concurrent_sync.get();
-    let mut next_tasks = Vec::new();
 
     // request the first task in blocking fashion to do less meaningless work
-    if let Some(first_task) = sync_queue::next_task(receiver).await {
-        next_tasks.push(first_task);
-    } else {
-        return LoopStep::Shutdown;
-    };
-    next_tasks.extend(
-        sync_queue::next_task_batch(receiver, max_concurrent_sync - 1)
-            .await
-            .into_iter(),
-    );
+    let (first_sync_id, first_task) =
+        if let Some(first_task) = sync_queue::next_task(receiver).await {
+            first_task
+        } else {
+            return ControlFlow::Break(());
+        };
+
+    let mut batched_tasks = sync_queue::next_task_batch(receiver, max_concurrent_sync - 1).await;
+    match batched_tasks.entry(first_sync_id) {
+        hash_map::Entry::Occupied(o) => {
+            let current = o.remove();
```

Comment:
`get_mut()` would require us to have a `fn merge(&mut self, ...)` method signature which did not work for me: the code that wants to replace `Upload`/`Download` with `DownloadAndUpload` fails due to borrow checker.

In theory, we could update such references as `*self = ...`, but since we also `match` on the reference before, the borrow checker is not happy. 
Additionally, with `&mut` reference, we have to `clone()` the data that we want to merge for the same `DownloadAndUpload` case.

So it feels that the method should be `fn merge(self, ...)` for now, before we untangle the `DownloadAndUpload` case first.